pubsub_logging/async_handler.py

```diff
@@ -0,0 +1,145 @@
+                 client=None, publish_body=publish_body, stderr_logger=None):
+        """The constructor of the handler.
+
+        Args:
+          topic: Cloud Pub/Sub topic name to send the logs.
+          worker_size: The initial worker size, defaults to 1.
+          retry: How many times to retry upon Cloud Pub/Sub API failure,
+                 defaults to 5.
+          timeout: Timeout for BatchQueue.get, defaults to 1.
+          client: An optional Cloud Pub/Sub client to use. If not set, one is
+                  built automatically, defaults to None.
+          publish_body: A callable for publishing the Pub/Sub message,
+                        just for testing purposes.
+          stderr_logger: A logger for informing failures with this
+                         logger, defaults to None and if not specified, a last
+                         resort logger will be used.
+        """
+        super(AsyncPubsubHandler, self).__init__()
+        self._topic = topic
+        self._retry = retry
+        self._client = client
+        self._q = mp.JoinableQueue()
+        self._should_exit = mp.Value('i', 0)
+        self._workers = []
+        self._timeout = timeout
+        self._threshold = BATCH_SIZE
+        self._buf = []
+        self._publish_body = publish_body
+        if stderr_logger:
+            self._stderr_logger = stderr_logger
+        else:
+            self._stderr_logger = logging.Logger('last_resort')
+            self._stderr_logger.addHandler(logging.StreamHandler())
+        for _ in range(worker_size):
+            p = mp.Process(target=self.send_loop, args=(self._q,))
+            p.daemon = True
+            self._workers.append(p)
+            p.start()
+
+    def send_loop(self, q):  # pragma: NO COVER
+        """Process loop for indefinitely sending logs to Cloud Pub/Sub."""
+        if self._client:
+            client = self._client
+        else:  # pragma: NO COVER
+            client = get_pubsub_client()
+        while not self._should_exit.value:
+            logs = []
+            try:
+                logs = q.get(block=True, timeout=self._timeout)
+            except Empty:
+                pass
+            if not logs:
+                continue
+            try:
+                body = {'messages':
+                        [{'data': compat_urlsafe_b64encode(self.format(r))}
+                            for r in logs]}
+                self._publish_body(client, body, self._topic, self._retry)
+            except errors.RecoverableError as e:
+                # Records the exception and puts the logs back to the deque
+                # and prints the exception to stderr.
+                q.put(logs)
+                self._stderr_logger.exception(e)
+            except Exception as e:
+                self._stderr_logger.exception(e)
+                self._stderr_logger.warn('There was a non recoverable error, '
+                                         'exiting.')
+                exit()
```

Comment:
Uh, I don't think `exit()` is available outside the context of the interactive interpreter.